compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -489,9 +489,15 @@ public void visit(BLangSimpleVarRef varRefExpr) {
         }
 
         // Check type compatibility
-        if (expType.tag == TypeTags.ARRAY && ((BArrayType) expType).state == BArrayState.OPEN_SEALED) {
-            dlog.error(varRefExpr.pos, DiagnosticCode.INVALID_USAGE_OF_SEALED_TYPE, "can not infer array size");
-            return;
+        if (expType.tag == TypeTags.ARRAY) {
+            BType type = expType;
+            do {
+                if (((BArrayType) type).state == BArrayState.OPEN_SEALED) {
+                    dlog.error(varRefExpr.pos, DiagnosticCode.INVALID_USAGE_OF_SEALED_TYPE, "can not infer array size");
+                    return;
+                }
+                type = ((BArrayType) type).eType;
+            } while (type.tag == TypeTags.ARRAY);
```

Comment:
Similar to above, we can write a recursive method, which i think would be simple and clean.